lgc/builder/BuilderImpl.cpp

```diff
@@ -329,6 +330,113 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+        : it(it), indexToInstr(indexToInstr) {}
+    iterator &operator++() {
+      ++it;
+      return *this;
+    }
+
+    Instruction *operator*() {
+      unsigned index = *it;
+      assert(index < indexToInstr.size() && "Index out of range.");
+      return indexToInstr[index];
+    }
+
+    bool operator!=(const iterator &otherIt) {
+      assert(&otherIt.indexToInstr == &indexToInstr && "Iterators of different objects.");
+      return otherIt.it != it;
+    }
+  };
+
+  iterator begin(const IndexToInstructionVec &indexToInstr) const {
+    return iterator(bits.set_bits_begin(), indexToInstr);
+  }
+
+  iterator end(const IndexToInstructionVec &indexToInstr) const { return iterator(bits.set_bits_end(), indexToInstr); }
+
+  void insert(Instruction *instr, const InstrToIndexMap &instrToIndex) {
+    auto it = instrToIndex.find(instr);
+    assert(it != instrToIndex.end() && "Expected to find instr in instrToIndex.");
+    unsigned index = it->second;
+    bits.set(index);
+  }
+
+  unsigned size() const { return bits.size(); }
+
+  bool empty() const { return bits.empty(); }
+};
+
+class TraceNonUniformIndex {
+  // Maps the instruction to its index in the bit vector.
+  TinyInstructionSet::InstrToIndexMap instrToIndex;
+  // The instructions used as keys in instrToIndex in program order. It is used to map an index to an instruction.
+  TinyInstructionSet::IndexToInstructionVec indexToInstr;
+  // Maps an instruction to its dependencies.
+  DenseMap<Value *, TinyInstructionSet> instrDeps;
+  bool scalarizeDescriptorLoads;
+
+  void insertNewValueInInsnDeps(Value *, Instruction *);
+
+public:
+  TraceNonUniformIndex(Instruction *nonUniformInst, bool scalarizeDescriptorLoads = false)
+      : scalarizeDescriptorLoads(scalarizeDescriptorLoads) {
+    // Initialization of instrToIndex and indexToInstr.
+    if (scalarizeDescriptorLoads) {
+      unsigned cnt = 0;
+      for (Instruction *I = nonUniformInst->getPrevNode(); I != nullptr && cnt < 64; I = I->getPrevNode(), ++cnt) {
+        indexToInstr.push_back(I);
+        instrToIndex[I] = cnt;
+      }
+    }
+  }
+
+  Value *run(Value *);
+
+  const DenseMap<Value *, TinyInstructionSet> &getInstrDeps() const { return instrDeps; }
+
+  const TinyInstructionSet::IndexToInstructionVec &getIndexToInstr() const { return indexToInstr; }
+
+  bool foundDependencies() const { return scalarizeDescriptorLoads; }
+};
+
+void TraceNonUniformIndex::insertNewValueInInsnDeps(Value *newValue, Instruction *currentVisitedInsn) {
+  if (!instrToIndex.contains(currentVisitedInsn)) {
+    // The instruction is either outside of 64 limit or in a different basic block. So, we bail-out scalarization.
+    scalarizeDescriptorLoads = false;
+    return;
+  }
+  assert(instrDeps.contains(currentVisitedInsn) && "The current visited instruction should have been in the map.");
+  auto it1 = instrDeps.try_emplace(newValue).first;
```

Comment:
Can you use a structured binding so this is a bit cleaner?

Reply:
Done.

Reply:
Yes, but you can check for the boolean value directly. However, it is just personal taste.

Reply:
I am not using the boolean. That's why I did not add it.